mmcls/runner/hooks/precise_bn.py

```diff
@@ -0,0 +1,177 @@
+        if isinstance(m, (torch.nn.BatchNorm2d, torch.nn.BatchNorm1d,
+                          torch.nn.BatchNorm3d))
+    ]
+
+    if len(bn_layers) == 0:
+        print_log('No BN found in model', logger=logger, level=logging.WARNING)
+        return
+    print_log(
+        f'{len(bn_layers)} BN found, run {num_iter} iters...', logger=logger)
+
+    # Initialize BN stats storage for computing
+    # mean(mean(batch)) and mean(var(batch))
+    running_means = [torch.zeros_like(bn.running_mean) for bn in bn_layers]
+    running_vars = [torch.zeros_like(bn.running_var) for bn in bn_layers]
+    # Remember momentum values
+    momentums = [bn.momentum for bn in bn_layers]
+    # Set momentum to 1.0 to compute BN stats that reflect the current batch
+    for bn in bn_layers:
+        bn.momentum = 1.0
+    # Average the BN stats for each BN layer over the batches
+    if rank == 0:
+        prog_bar = mmcv.ProgressBar(num_iter)
+
+    count = 0
+    finish = False
+    for loader in loaders:
+        for data in loader:
+            parallel_module(**data)
+            for i, bn in enumerate(bn_layers):
+                running_means[i] += bn.running_mean / num_iter
+                running_vars[i] += bn.running_var / num_iter
+            count += 1
+            if count >= num_iter:
+                finish = True
+                break
+
+            if rank == 0:
+                prog_bar.update()
+        if finish:
+            break
+
+    # Sync BN stats across GPUs (no reduction if 1 GPU used)
+    running_means = scaled_all_reduce(running_means, NUM_GPUS)
+    running_vars = scaled_all_reduce(running_vars, NUM_GPUS)
+    # Set BN stats and restore original momentum values
+    for i, bn in enumerate(bn_layers):
+        bn.running_mean = running_means[i]
+        bn.running_var = running_vars[i]
+        bn.momentum = momentums[i]
+
+
+class PreciseBNHook(Hook):
+    """Precise BN hook.
+
+    Attributes:
+        dataloaders (DataLoader): A List of PyTorch dataloader.
+        num_items (int): Number of iterations to update the bn stats.
+            Default: 8192.
+        interval (int): Perform precise bn interval (by epochs). Default: 1.
+    """
+
+    def __init__(self, dataloaders, num_items=8192, interval=1):
+        assert len(dataloaders) > 0, 'dataloaders is empty...'
+        if not isinstance(dataloaders, list):
+            raise TypeError('dataloaders must be a List,but got',
+                            f' {type(dataloaders)}')
+        if not isinstance(dataloaders[0], DataLoader):
+            raise TypeError(
+                'dataloaders must be a list of Pytorch Dataloaders ,but got',
+                f' {type(dataloaders[0])}')
+        self.dataloaders = dataloaders
+        self.interval = interval
+        self.num_items = num_items
+
+    def after_train_epoch(self, runner):
+        if self.every_n_epochs(runner, self.interval):
+            # sleep to avoid possible deadlock
```

Comment:
possible deadlock?

Reply:
I refer to the code of [mmaction2](https://github.com/open-mmlab/mmaction2/blob/075464c9157a20a33ccecf11ed3a532ecfbb3b78/mmaction/utils/precise_bn.py#L144). 

There should be no risk of deadlock here. 
Both two "sleep" lines can be deleted here.